packages/lib/services/joplinCloudUtils.ts

```diff
@@ -0,0 +1,112 @@
+	next: ActionType;
+	active: ActionType | 'INITIAL';
+	errorMessage?: string;
+};
+
+export const defaultState: DefaultState = {
+	className: 'text',
+	message: ()=> _('Waiting for authorisation...'),
+	next: 'LINK_USED',
+	active: 'INITIAL',
+};
+
+export const reducer: Reducer<DefaultState, Action> = (state: DefaultState, action: Action) => {
+	switch (action.type) {
+	case 'LINK_USED': {
+		return {
+			className: 'text',
+			message: () => _('If you have already authorised, please wait for the application to sync to Joplin Cloud.'),
+			next: 'COMPLETED',
+			active: 'LINK_USED',
+		};
+	}
+	case 'COMPLETED': {
+		return {
+			className: 'bold',
+			message: () => _('You are logged in into Joplin Cloud, you can leave this screen now.'),
+			active: 'COMPLETED',
+			next: 'COMPLETED',
+		};
+	}
+	case 'ERROR': {
+		return {
+			className: 'text',
+			message: () => _('You were unable to connect to Joplin Cloud, verify your connection. Error: '),
+			active: 'ERROR',
+			next: 'COMPLETED',
+			errorMessage: action.payload,
+		};
+	}
+	default: {
+		return state;
+	}
+	}
+};
+
+export const getApplicationInformation = async () => {
+	const platformName = await shim.platformName();
+	switch (platformName) {
+	case 'ios':
+		return { type: ApplicationType.Mobile, platform: ApplicationPlatform.Ios };
+	case 'android':
+		return { type: ApplicationType.Mobile, platform: ApplicationPlatform.Android };
+	case 'darwin':
+		return { type: ApplicationType.Desktop, platform: ApplicationPlatform.MacOs };
+	case 'win32':
+		return { type: ApplicationType.Desktop, platform: ApplicationPlatform.Windows };
+	case 'linux':
+		return { type: ApplicationType.Desktop, platform: ApplicationPlatform.Linux };
+	default:
+		return { type: ApplicationType.Unknown, platform: ApplicationPlatform.Unknown };
+	}
+};
+
+export const generateLoginWithUniqueLoginCode = async (loginUrl: string, uniqueloginCode: string) => {
+	const applicationInfo = await getApplicationInformation();
+	const searchParams = new URLSearchParams();
+	searchParams.append('unique_login_code', uniqueloginCode);
+	searchParams.append('platform', applicationInfo.platform.toString());
+	searchParams.append('type', applicationInfo.type.toString());
+	searchParams.append('version', shim.appVersion());
+
+	return `${loginUrl}?${searchParams.toString()}`;
+};
+
+let isWaitingResponse = false;
+
+export const checkIfLoginWasSuccessful = async (applicationsUrl: string, ulc: string) => {
+	if (isWaitingResponse) return undefined;
+	isWaitingResponse = true;
+
+	const response = await fetch(`${applicationsUrl}?unique_login_code=${ulc}`);
+
+	if (response.ok) {
+		const jsonResponse = await response.json();
+		Setting.setValue('sync.10.username', jsonResponse.id);
+		Setting.setValue('sync.10.password', jsonResponse.password);
+		return { success: true };
+	}
+
+	isWaitingResponse = false;
+	const jsonBody = await response.json();
+	logger.warn('Server could not retrieve application credential', jsonBody);
```

Comment:
Does it mean that when trying to fetch `unique_login_code` but it's not ready yet the server throws an error? If that's what it is I think it should be changed - instead it should tell the status like "in progress", "done", etc. with 200 status code.

Errors should be only when there's an actual error, for example if the code doesn't exist or the server is down. And we need to report these errors to the user so that function should throw an exception with the statusText or whatever info the server provides.

Reply:
I made the change to follow this principle. I wasn't sure if it was ok to keep the message log, I guess it doesn't make sense from what you saying.